File: frontend/src/components/act/vocabulary.ts
// ── Closed vocabularies for the EQ2Parser-enrichment tag fields ──────────────
//
// Mirrors backend/server/api/act/vocabulary.py (which mirrors EQ2Parser's
// Core Vocabulary). Damage schools are spelled the way the combat log spells
// them; control effects use tooltip vocabulary ("stifle", not the log's
// "silenced"). These never appear in ACT XML exports — they ride the
// /api/act/pack JSON for EQ2Parser users.

export const DAMAGE_SCHOOLS = [
  'crushing',
  'slashing',
  'piercing',
  'heat',
  'cold',
  'magic',
  'mental',
  'divine',
  'disease',
  'poison',
  'focus',
] as const

export const CONTROL_EFFECTS = [
  'stun',
  'stifle',
  'mez',
  'fear',
  'root',
  'snare',
  'daze',
  'charm',
  'knockup',
] as const

/** "poison, disease" → ["poison", "disease"] (order = dominance). */
export function parseSchools(value: string): string[] {
  return value
    .split(',')
    .map(s => s.trim().toLowerCase())
    .filter(s => (DAMAGE_SCHOOLS as readonly string[]).includes(s))
}

/** Toggle one school in the comma-joined list. Newly toggled schools append,
 * so toggle ORDER is dominance order — first picked leads (it drives the
 * parser's flame colour). */
export function toggleSchool(value: string, school: string): string {
  const schools = parseSchools(value)
  const next = schools.includes(school)
    ? schools.filter(s => s !== school)
    : [...schools, school]
  return next.join(', ')
}

File: backend/server/api/act/vocabulary.py
"""Closed vocabularies for the EQ2Parser-enrichment tag fields.

Mirrors EQ2Parser's ``EQ2Parser.Core.Triggers.Vocabulary`` — the parser,
the pack endpoint, and the web editors must agree on these or tags drift.
Damage schools are spelled the way the combat log spells them; control
effects use tooltip vocabulary (the log says "silenced", tooltips say
Stifle — curators speak tooltip).
"""

from __future__ import annotations

DAMAGE_SCHOOLS: tuple[str, ...] = (
    "crushing",
    "slashing",
    "piercing",
    "heat",
    "cold",
    "magic",
    "mental",
    "divine",
    "disease",
    "poison",
    "focus",
)

CONTROL_EFFECTS: tuple[str, ...] = (
    "stun",
    "stifle",
    "mez",
    "fear",
    "root",
    "snare",
    "daze",
    "charm",
    "knockup",
)


def normalize_damage_type(value: str) -> str:
    """Validate + canonicalise a comma-joined school list, preserving the
    caller's order (dominant school first — it drives the parser's flame
    colour). Unknown tokens raise ValueError → Pydantic 422."""
    seen: list[str] = []
    for token in value.split(","):
        school = token.strip().lower()
        if not school:
            continue
        if school not in DAMAGE_SCHOOLS:
            raise ValueError(f"unknown damage school {school!r} (allowed: {', '.join(DAMAGE_SCHOOLS)})")
        if school not in seen:
            seen.append(school)
    return ", ".join(seen)


def normalize_control_effect(value: str) -> str:
    effect = value.strip().lower()
    if effect and effect not in CONTROL_EFFECTS:
        raise ValueError(f"unknown control effect {effect!r} (allowed: {', '.join(CONTROL_EFFECTS)})")
    return effect
